### PR TITLE
Propagation API

### PR DESCRIPTION
**Pull request by EddeCCC**
Add an exporter that contains a REST-API running on a HTTP-Server inside the Ocelot agent. The REST-API allows to export and also import propagation data.

- - -
This change is [Reviewable](https://reviewable.io/reviews/inspectIT/inspectit-ocelot/1615)

### CODE REVIEW

**Review by quandor — CHANGES_REQUESTED** [review]
> Reviewed 12 of 15 files at r1, 8 of 8 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/inspectIT/inspectit-ocelot/1615)* status: all files reviewed, 5 unresolved discussions (waiting on @EddeCCC)

___
*[`inspectit-ocelot-core/build.gradle` line 78 at r2](https://reviewable.io/reviews/inspectIT/inspectit-ocelot/1615#-NaFsqjZBQXAmra5g35p:-NaFsqjZBQXAmra5g35q:b-bfwacc) ([raw file](https://github.com/inspectIT/inspectit-ocelot/blob/4b13b2b4074d385403c7ed8001663486f641bef6/inspectit-ocelot-core/build.gradle#L78)):*
> ```Gradle
> 
>             // jetty HTTP server
>             'org.eclipse.jetty:jetty-server:9.4.51.v20230217',
> ```

I am not sure whether here would be the right place. But it might make sense to note down the rationale why  Jetty was used instead of making the agent a full blown Spring-Boot App. You mentioned low footprint to me. Another reason _might_ be that it is not easy to stop and start the HTTP part of a full blown Spring-Boot App .
___
*[`inspectit-ocelot-core/src/main/java/rocks/inspectit/ocelot/core/exporter/BrowserPropagationServlet.java` line 30 at r2](https://reviewable.io/reviews/inspectIT/inspectit-ocelot/1615#-NaFtSQz7X87g3vCts-z:-NaFtSQz7X87g3vCts0-:bwvz10l) ([raw file](https://github.com/inspectIT/inspectit-ocelot/blob/4b13b2b4074d385403c7ed8001663486f641bef6/inspectit-ocelot-core/src/main/java/rocks/inspectit/ocelot/core/exporter/BrowserPropagationServlet.java#L30)):*
> ```Java
>     private final ObjectMapper mapper;
>     private final BrowserPropagationDataStorage dataStorage;
>     private int responseCode;
> ```

I would try to avoid this member variable. I will write some comments further down on how and why to do that.
___
*[`inspectit-ocelot-core/src/main/java/rocks/inspectit/ocelot/core/exporter/BrowserPropagationServlet.java` line 44 at r2](https://reviewable.io/reviews/inspectIT/inspectit-ocelot/1615#-NaFtcIz0Ssxe_oI6zp-:-NaFtcIz0Ssxe_oI6zp0:bim504c) ([raw file](https://github.com/inspectIT/inspectit-ocelot/blob/4b13b2b4074d385403c7ed8001663486f641bef6/inspectit-ocelot-core/src/main/java/rocks/inspectit/ocelot/core/exporter/BrowserPropagationServlet.java#L44)):*
> ```Java
>     public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
>         log.info("Tags HTTP-server received GET-request");
>         setResponseCode(HttpServletResponse.SC_OK);
> ```

Instead of calling this setter, just use the constant a few lines below in `response.setStatus(..)`. I do not see any advantage in using a member variable.
___
*[`inspectit-ocelot-core/src/main/java/rocks/inspectit/ocelot/core/exporter/BrowserPropagationServlet.java` line 70 at r2](https://reviewable.io/reviews/inspectIT/inspectit-ocelot/1615#-NaFtrgSAWjAabsVHaLC:-NaFtrgSAWjAabsVHaLD:bscnpye) ([raw file](https://github.com/inspectIT/inspectit-ocelot/blob/4b13b2b4074d385403c7ed8001663486f641bef6/inspectit-ocelot-core/src/main/java/rocks/inspectit/ocelot/core/exporter/BrowserPropagationServlet.java#L70)):*
> ```Java
>                     .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
>         } catch (Exception e) {
>             log.info("Request failed");
> ```

If you would handle that Exception in `doPut` you will not need the member variable. By doing so you achieve IMHO two things:
* You finally can get rid of the member variable and thus making the code less complex
* Response code handling is "at the same level" in the class. Currently you are handling response code in `get`, but `put` hands it of to a method further "down". Your code gets more symmetric if response code handling is in `get` and `put`. Basically you have a something like a template used in both methods. Get a request, handle that, set the response code. That helps IMHO understanding and remembering the code.
___
*[`inspectit-ocelot-core/src/test/java/rocks/inspectit/ocelot/core/exporter/BrowserPropagationHttpExporterServiceIntTest.java` line 36 at r2](https://reviewable.io/reviews/inspectIT/inspectit-ocelot/1615#-NaFvBL3AlW8Nn0bOFtH:-NaFvBL3AlW8Nn0bOFtI:b-naex51) ([raw file](https://github.com/inspectIT/inspectit-ocelot/blob/4b13b2b4074d385403c7ed8001663486f641bef6/inspectit-ocelot-core/src/test/java/rocks/inspectit/ocelot/core/exporter/BrowserPropagationHttpExporterServiceIntTest.java#L36)):*
> ```Java
>     BrowserPropagationDataStorage dataStorage;
> 
>     private static CloseableHttpClient testClient;
> ```

Why is `testClient`static if you recreate it before every test anyway? Either create it only once or remove that `static`.
___
*[`inspectit-ocelot-core/src/test/java/rocks/inspectit/ocelot/core/exporter/BrowserPropagationHttpExporterServiceIntTest.java` line 45 at r2](https://reviewable.io/reviews/inspectIT/inspectit-ocelot/1615#-NaFvUmR74WFFeyq-IWw:-NaFvUmR74WFFeyq-IWx:b-f5kkvu) ([raw file](https://github.com/inspectIT/inspectit-ocelot/blob/4b13b2b4074d385403c7ed8001663486f641bef6/inspectit-ocelot-core/src/test/java/rocks/inspectit/ocelot/core/exporter/BrowserPropagationHttpExporterServiceIntTest.java#L45)):*
> ```Java
>     void prepareTest() throws IOException {
>         startServer();
>         createHttpClient();
> ```

I would prefer `createHttpClient` returning a ClosableHttpClient and setting it here. For reading code I more easily grasp that this method simply creates a Client to be used later on. Currently I can only assume that and would need to look at it verifying my assumption.

**Review by TitusLabs — CHANGES_REQUESTED** [review]

File: inspectit-ocelot-core/src/main/java/rocks/inspectit/ocelot/core/instrumentation/browser/BrowserPropagationDataStorage.java [hunk 1]
```diff
@@ -8,8 +10,11 @@
  *  DataStorage for all tags, that should be propagated to one browser
  *  Normally, there should be only one data storage per session
  */
+@Slf4j
 public class BrowserPropagationDataStorage {
```
Comment (TitusLabs) [inline inspectit-ocelot-core/src/main/java/rocks/inspectit/ocelot/core/instrumentation/browser/BrowserPropagationDataStorage.java:14]: Mir Fehlt noch die Größenbegrenzung einer Session.

File: inspectit-ocelot-core/src/main/java/rocks/inspectit/ocelot/core/instrumentation/browser/BrowserPropagationDataStorage.java [hunk 2]
```diff
@@ -38,4 +47,8 @@ public void updateTimestamp(long newTimestamp) {
     public long calculateElapsedTime(long currentTime) {
         return currentTime - latestTimestamp;
     }
+
+    private boolean validateAttributeLength(Map<String, Object> newPropagationData) {
+        return propagationData.size() + newPropagationData.size() <= ATTRIBUTE_COUNT_LIMIT;
```
Comment (TitusLabs) [inline inspectit-ocelot-core/src/main/java/rocks/inspectit/ocelot/core/instrumentation/browser/BrowserPropagationDataStorage.java:52]: Dies berücksichtigt nicht den Fall, dass bestehende Werte überschrieben werden - wollen wir das?

File: inspectit-ocelot-config/src/main/resources/rocks/inspectit/ocelot/config/default/exporters.yml [hunk 3]
```diff
@@ -111,5 +111,7 @@ inspectit:
         port: 9000
         # the path for the endpoint of the http-server
         path: "/inspectit"
+        # how many sessions can be stored at the same time
+        session-limit: 32
```
Comment (TitusLabs) [inline inspectit-ocelot-config/src/main/resources/rocks/inspectit/ocelot/config/default/exporters.yml:115]: Der Wert dürfte etwas zu klein sein..

**Review by TitusLabs — CHANGES_REQUESTED** [review]

File: inspectit-ocelot-core/src/main/java/rocks/inspectit/ocelot/core/instrumentation/browser/BrowserPropagationDataStorage.java (hunk 1 above)
Comment (EddeCCC) [reply inspectit-ocelot-core/src/main/java/rocks/inspectit/ocelot/core/instrumentation/browser/BrowserPropagationDataStorage.java:14]: Ich hätte gedacht, dass ich durch die Variable `ATTRIBUTE_COUNT_LIMIT` eine Session begrenze. Oder was meinst du?

File: inspectit-ocelot-config/src/main/resources/rocks/inspectit/ocelot/config/default/exporters.yml (hunk 3 above)
Comment (EddeCCC) [reply inspectit-ocelot-config/src/main/resources/rocks/inspectit/ocelot/config/default/exporters.yml:115]: Ist nur ein default-Wert denn man überschreiben kann. Was hättest du im Sinne^^? 100?

**Review by EddeCCC — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/inspectIT/inspectit-ocelot/1615)* status: 9 of 25 files reviewed, 16 unresolved discussions (waiting on @quandor and @TitusLabs)

___
*[`inspectit-ocelot-core/build.gradle` line 78 at r2](https://reviewable.io/reviews/inspectIT/inspectit-ocelot/1615#-NaFsqjZBQXAmra5g35p:-NaowysX87QZ9H_5sf3w:b-896fix) ([raw file](https://github.com/inspectIT/inspectit-ocelot/blob/4b13b2b4074d385403c7ed8001663486f641bef6/inspectit-ocelot-core/build.gradle#L78)):*
<details><summary><i>Previously, quandor (Jochen Just) wrote…</i></summary><blockquote>

I am not sure whether here would be the right place. But it might make sense to note down the rationale why  Jetty was used instead of making the agent a full blown Spring-Boot App. You mentioned low footprint to me. Another reason _might_ be that it is not easy to stop and start the HTTP part of a full blown Spring-Boot App .
</blockquote></details>

Done.
___
*[`inspectit-ocelot-core/build.gradle` line 81 at r3](https://reviewable.io/reviews/inspectIT/inspectit-ocelot/1615#gh-1279122771:-NaowzHfAfl-8cLNfq63:b-896fix) ([raw file](https://github.com/inspectIT/inspectit-ocelot/blob/3def14b6face9731cec2bd4551a92528ccc8e3b6/inspectit-ocelot-core/build.gradle#L81)):*
<details><summary><i>Previously, TitusLabs (Titus Meyer) wrote…</i></summary><blockquote>

Die Versionen bitte noch auslagern.
</blockquote></details>

Done.
___
*[`inspectit-ocelot-core/src/main/java/rocks/inspectit/ocelot/core/exporter/BrowserPropagationHttpExporterService.java` line 73 at r3](https://reviewable.io/reviews/inspectIT/inspectit-ocelot/1615#gh-1279132548:-Naox-QXAXeyIhbtQ3_U:b-896fix) ([raw file](https://github.com/inspectIT/inspectit-ocelot/blob/3def14b6face9731cec2bd4551a92528ccc8e3b6/inspectit-ocelot-core/src/main/java/rocks/inspectit/ocelot/core/exporter/BrowserPropagationHttpExporterService.java#L73)):*
<details><summary><i>Previously, TitusLabs (Titus Meyer) wrote…</i></summary><blockquote>

Das Log vor der Aktion schreiben.
</blockquote></details>

Done.
___
*[`inspectit-ocelot-core/src/main/java/rocks/inspectit/ocelot/core/exporter/BrowserPropagationHttpExporterService.java` line 90 at r3](https://reviewable.io/reviews/inspectIT/inspectit-ocelot/1615#gh-1279132784:-Naox-je5UYCXqi8czwd:b-896fix) ([raw file](https://github.com/inspectIT/inspectit-ocelot/blob/3def14b6face9731cec2bd4551a92528ccc8e3b6/inspectit-ocelot-core/src/main/java/rocks/inspectit/ocelot/core/exporter/BrowserPropagationHttpExporterService.java#L90)):*
<details><summary><i>Previously, TitusLabs (Titus Meyer) wrote…</i></summary><blockquote>

Hier auch vor der Aktion.
</blockquote></details>

Done.
___
*[`inspectit-ocelot-core/src/main/java/rocks/inspectit/ocelot/core/exporter/BrowserPropagationHttpExporterService.java` line 104 at r3](https://reviewable.io/reviews/inspectIT/inspectit-ocelot/1615#gh-1279134468:-Naox0SRAq7UhYIdBfON:b-896fix) ([raw file](https://github.com/inspectIT/inspectit-ocelot/blob/3def14b6face9731cec2bd4551a92528ccc8e3b6/inspectit-ocelot-core/src/main/java/rocks/inspectit/ocelot/core/exporter/BrowserPropagationHttpExporterService.java#L104)):*
<details><summary><i>Previously, TitusLabs (Titus Meyer) wrote…</i></summary><blockquote>

Es ist ok, wenn du in diesem Falle nichts unternimmst. Allerdings musst du dann auch in doDisable() den Cache löschen.
</blockquote></details>

Done.
___
*[`inspectit-ocelot-core/src/main/java/rocks/inspectit/ocelot/core/exporter/BrowserPropagationServlet.java` line 30 at r2](https://reviewable.io/reviews/inspectIT/inspectit-ocelot/1615#-NaFtSQz7X87g3vCts-z:-Naox1NI8ZRXWfra4weg:b-896fix) ([raw file](https://github.com/inspectIT/inspectit-ocelot/blob/4b13b2b4074d385403c7ed8001663486f641bef6/inspectit-ocelot-core/src/main/java/rocks/inspectit/ocelot/core/exporter/BrowserPropagationServlet.java#L30)):*
<details><summary><i>Previously, quandor (Jochen Just) wrote…</i></summary><blockquote>

I would try to avoid this member variable. I will write some comments further down on how and why to do that.
</blockquote></details>

Done.
___
*[`inspectit-ocelot-core/src/main/java/rocks/inspectit/ocelot/core/exporter/BrowserPropagationServlet.java` line 44 at r2](https://reviewable.io/reviews/inspectIT/inspectit-ocelot/1615#-NaFtcIz0Ssxe_oI6zp-:-Naox4kXAepn_lByd9Kq:b-896fix) ([raw file](https://github.com/inspectIT/inspectit-ocelot/blob/4b13b2b4074d385403c7ed8001663486f641bef6/inspectit-ocelot-core/src/main/java/rocks/inspectit/ocelot/core/exporter/BrowserPropagationServlet.java#L44)):*
<details><summary><i>Previously, quandor (Jochen Just) wrote…</i></summary><blockquote>

Instead of calling this setter, just use the constant a few lines below in `response.setStatus(..)`. I do not see any advantage in using a member variable.
</blockquote></details>

Done.
___
*[`inspectit-ocelot-core/src/main/java/rocks/inspectit/ocelot/core/exporter/BrowserPropagationServlet.java` line 70 at r2](https://reviewable.io/reviews/inspectIT/inspectit-ocelot/1615#-NaFtrgSAWjAabsVHaLC:-Naox6cP6NkI6bHlhBwJ:b-896fix) ([raw file](https://github.com/inspectIT/inspectit-ocelot/blob/4b13b2b4074d385403c7ed8001663486f641bef6/inspectit-ocelot-core/src/main/java/rocks/inspectit/ocelot/core/exporter/BrowserPropagationServlet.java#L70)):*
<details><summary><i>Previously, quandor (Jochen Just) wrote…</i></summary><blockquote>

If you would handle that Exception in `doPut` you will not need the member variable. By doing so you achieve IMHO two things:
* You finally can get rid of the member variable and thus making the code less complex
* Response code handling is "at the same level" in the class. Currently you are handling response code in `get`, but `put` hands it of to a method further "down". Your code gets more symmetric if response code handling is in `get` and `put`. Basically you have a something like a template used in both methods. Get a request, handle that, set the response code. That helps IMHO understanding and remembering the code.
</blockquote></details>

Done.
___
*[`inspectit-ocelot-core/src/main/java/rocks/inspectit/ocelot/core/exporter/BrowserPropagationServlet.java` line 39 at r3](https://reviewable.io/reviews/inspectIT/inspectit-ocelot/1615#gh-1279183829:-Naox2Am5hbSs8Suv6-4:b-896fix) ([raw file](https://github.com/inspectIT/inspectit-ocelot/blob/3def14b6face9731cec2bd4551a92528ccc8e3b6/inspectit-ocelot-core/src/main/java/rocks/inspectit/ocelot/core/exporter/BrowserPropagationServlet.java#L39)):*
<details><summary><i>Previously, TitusLabs (Titus Meyer) wrote…</i></summary><blockquote>

Hier würde ich nicht höher, als Debug gehen.
</blockquote></details>

Done.
___
*[`inspectit-ocelot-core/src/main/java/rocks/inspectit/ocelot/core/exporter/BrowserPropagationServlet.java` line 64 at r3](https://reviewable.io/reviews/inspectIT/inspectit-ocelot/1615#gh-1279186417:-Naox5Kw8V4PHf9ABT6v:b-896fix) ([raw file](https://github.com/inspectIT/inspectit-ocelot/blob/3def14b6face9731cec2bd4551a92528ccc8e3b6/inspectit-ocelot-core/src/main/java/rocks/inspectit/ocelot/core/exporter/BrowserPropagationServlet.java#L64)):*
<details><summary><i>Previously, TitusLabs (Titus Meyer) wrote…</i></summary><blockquote>

Die Methode ist schon sehr gefährlich. Mal unabhängig davon, dass man nicht Massen an neuen Sessions erstellt, so muss wenigsten die Anzahl und die Größe der zu speichernden Attribute pro Session eingeschränkt werden. Zur Not noch eine Obergrenze für Session-Einträge definieren...
</blockquote></details>

Done.
___
*[`inspectit-ocelot-core/src/main/java/rocks/inspectit/ocelot/core/exporter/BrowserPropagationServlet.java` line 65 at r3](https://reviewable.io/reviews/inspectIT/inspectit-ocelot/1615#gh-1279183941:-Naox5tH6xdoQ-mLj2fc:b-896fix) ([raw file](https://github.com/inspectIT/inspectit-ocelot/blob/3def14b6face9731cec2bd4551a92528ccc8e3b6/inspectit-ocelot-core/src/main/java/rocks/inspectit/ocelot/core/exporter/BrowserPropagationServlet.java#L65)):*
<details><summary><i>Previously, TitusLabs (Titus Meyer) wrote…</i></summary><blockquote>

...ebenso
</blockquote></details>

Done.
___
*[`inspectit-ocelot-core/src/main/java/rocks/inspectit/ocelot/core/instrumentation/browser/BrowserPropagationDataStorage.java` line 52 at r4](https://reviewable.io/reviews/inspectIT/inspectit-ocelot/1615#gh-1280229860:-Naox8z2-s3NFkJuBE8u:b-896fix) ([raw file](https://github.com/inspectIT/inspectit-ocelot/blob/08edfdceaae6df5bb1080005d2875f1beaf78563/inspectit-ocelot-core/src/main/java/rocks/inspectit/ocelot/core/instrumentation/browser/BrowserPropagationDataStorage.java#L52)):*
<details><summary><i>Previously, TitusLabs (Titus Meyer) wrote…</i></summary><blockquote>

Dies berücksichtigt nicht den Fall, dass bestehende Werte überschrieben werden - wollen wir das?
</blockquote></details>

Done.
___
*[`inspectit-ocelot-core/src/main/java/rocks/inspectit/ocelot/core/instrumentation/browser/BrowserPropagationSessionStorage.java` line 29 at r3](https://reviewable.io/reviews/inspectIT/inspectit-ocelot/1615#gh-1279204889:-Naox9lX9lU3or9Db_JK:b-896fix) ([raw file](https://github.com/inspectIT/inspectit-ocelot/blob/3def14b6face9731cec2bd4551a92528ccc8e3b6/inspectit-ocelot-core/src/main/java/rocks/inspectit/ocelot/core/instrumentation/browser/BrowserPropagationSessionStorage.java#L29)):*
<details><summary><i>Previously, TitusLabs (Titus Meyer) wrote…</i></summary><blockquote>

computeIfAbsent() könnte es etwas vereinfachen
</blockquote></details>

Done.
___
*[`inspectit-ocelot-core/src/test/java/rocks/inspectit/ocelot/core/exporter/BrowserPropagationHttpExporterServiceIntTest.java` line 36 at r2](https://reviewable.io/reviews/inspectIT/inspectit-ocelot/1615#-NaFvBL3AlW8Nn0bOFtH:-NaoxAjn-kkkvOfWtWzx:b-896fix) ([raw file](https://github.com/inspectIT/inspectit-ocelot/blob/4b13b2b4074d385403c7ed8001663486f641bef6/inspectit-ocelot-core/src/test/java/rocks/inspectit/ocelot/core/exporter/BrowserPropagationHttpExporterServiceIntTest.java#L36)):*
<details><summary><i>Previously, quandor (Jochen Just) wrote…</i></summary><blockquote>

Why is `testClient`static if you recreate it before every test anyway? Either create it only once or remove that `static`.
</blockquote></details>

Done.

**Review by TitusLabs — APPROVED** [review]

**Review by TitusLabs — COMMENTED** [review]
> Reviewed 8 of 15 files at r1, 1 of 8 files at r2, 2 of 12 files at r3, 5 of 13 files at r4, 9 of 9 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/inspectIT/inspectit-ocelot/1615)* status: all files reviewed, 16 unresolved discussions (waiting on @quandor)

**Review by TitusLabs — COMMENTED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/inspectIT/inspectit-ocelot/1615#-:-NaqCcUo17VU-PPDQP6q:bnfp4nl)* status: all files reviewed, 7 unresolved discussions (waiting on @quandor)

**Review by EddeCCC — COMMENTED** [review]
> Reviewed 8 of 15 files at r1, 1 of 8 files at r2, 2 of 12 files at r3, 5 of 13 files at r4, 9 of 9 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/inspectIT/inspectit-ocelot/1615)* status: all files reviewed, 7 unresolved discussions (waiting on @quandor and @TitusLabs)

**Review by EddeCCC — COMMENTED** [review]
> Dismissed @quandor from 5 discussions.
> *[Reviewable](https://reviewable.io/reviews/inspectIT/inspectit-ocelot/1615)* status: all files reviewed, 2 unresolved discussions (waiting on @TitusLabs)

**Review by TitusLabs — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/inspectIT/inspectit-ocelot/1615)* status: all files reviewed, 1 unresolved discussion (waiting on @EddeCCC)

___
*[`inspectit-ocelot-core/src/main/java/rocks/inspectit/ocelot/core/instrumentation/browser/BrowserPropagationDataStorage.java` line 14 at r4](https://reviewable.io/reviews/inspectIT/inspectit-ocelot/1615#gh-1280229051:-NaqDeXp4W1Qz8ZLB96U:bak4tjl) ([raw file](https://github.com/inspectIT/inspectit-ocelot/blob/08edfdceaae6df5bb1080005d2875f1beaf78563/inspectit-ocelot-core/src/main/java/rocks/inspectit/ocelot/core/instrumentation/browser/BrowserPropagationDataStorage.java#L14)):*
<details><summary><i>Previously, EddeCCC (Eduard Schander) wrote…</i></summary><blockquote>

Ich hätte gedacht, dass ich durch die Variable `ATTRIBUTE_COUNT_LIMIT` eine Session begrenze. Oder was meinst du?
</blockquote></details>

wurde behoben

**Review by TitusLabs — APPROVED** [review]

**Review by Dimi-Ma — APPROVED** [review]

**Pull request merged**